### PR TITLE
DAOS-2189 vos: Direct Key Hash when Checksum Enabled

### PR DESCRIPTION
**Pull request by ryon-jensen**
Because the key hash function will be used for keys,
keys direct keys need to store the hash as well in
addition to the node

Signed-off-by: Ryon Jensen <ryon.jensen@intel.com>

### CODE REVIEW

File: src/common/btree.c [hunk 1]
```diff
@@ -392,6 +400,9 @@ btr_hkey_size_const(btr_ops_t *ops, uint64_t feats)
 {
 	int size;
 
+	if (BTR_IS_DIRECT_KEY(feats) && BTR_IS_CSUM_ENABLED(feats))
+		return sizeof(umem_off_t) + ops->to_hkey_size();
```
Comment (jolivier23) [inline src/common/btree.c]: you could avoid checking BTR_IS_DIRECT_KEY twice

if (BTR_IS_DIRECT_KEY(feats)) {
     if (BTR_IS_CSUM_ENABLED(feats))
           return ....
    else
          return ...
}

File: src/include/daos/btree.h [hunk 2]
```diff
@@ -33,6 +33,14 @@
 #include <daos_types.h>
 #include <daos/mem.h>
 
+/**
+ * With Checksum enabled, direct key should still have a hash
+ */
+struct btr_hash_direct {
+	umem_off_t		rec_node; /* direct key */
+	char		rec_hkey[0]; /* hashed key */
```
Comment (jolivier23) [inline src/include/daos/btree.h]: is the spacing off here or is it just github?

File: src/common/btree.c (hunk 1 above)
Comment (ryon-jensen) [reply src/common/btree.c]: changed

File: src/include/daos/btree.h (hunk 2 above)
Comment (ryon-jensen) [reply src/include/daos/btree.h]: fixed

**Comment by ryon-jensen** [comment]
Testing was done manually. I added an option to the btree_direct test that "enables" checksums.

File: src/common/btree.c [hunk 3]
```diff
@@ -1102,7 +1113,11 @@ btr_node_split_and_insert(struct btr_context *tcx, struct btr_trace *trace,
 		/* insert the right node and the first key of the right
 		 * node to its parent
 		 */
-		if (btr_is_direct_key(tcx))
+		if (btr_is_direct_key(tcx) && btr_is_csum_enabled(tcx)) {
+			rec->rec_hnode->rec_node = off_right;
+			btr_hkey_copy(tcx, &rec->rec_hnode->rec_hkey[0],
+				      &rec_dst->rec_hnode->rec_hkey[0]);
+		} else if (btr_is_direct_key(tcx))
```
Comment (jolivier23) [inline src/common/btree.c]: can check btr_is_direct_key once if you move the csum specific stuff into the if block

File: src/common/btree.c [hunk 4]
```diff
@@ -2024,8 +2055,16 @@ btr_node_del_leaf_rebal(struct btr_context *tcx,
 		par_rec = btr_node_rec_at(tcx, par_tr->tr_node,
 					  par_tr->tr_at - 1);
 		/* NB: Direct key of parent already points to this leaf */
-		if (!btr_is_direct_key(tcx))
-			btr_rec_copy_hkey(tcx, par_rec, dst_rec);
+		if (btr_is_direct_key(tcx) && btr_is_csum_enabled(tcx)) {
```
Comment (jolivier23) [inline src/common/btree.c]: same comment about checking btr_is_direct_key

File: src/common/btree.c (hunk 3 above)
Comment (gnailzenh) [reply src/common/btree.c]: indeed

**Comment by ryon-jensen** [comment]
Change not relevant anymore

**Pull request closed without merging**